File: services/ingestion-replication-service/src/domain/agent_registry.rs
use chrono::Utc;
use uuid::Uuid;

use crate::{AppState, models::agent::ConnectorAgent};

pub async fn resolve_agent_url(
    state: &AppState,
    connection_config: &serde_json::Value,
) -> Result<Option<String>, String> {
    if let Some(url) = connection_config
        .get("agent_url")
        .and_then(serde_json::Value::as_str)
    {
        let trimmed = url.trim();
        if !trimmed.is_empty() {
            return Ok(Some(trimmed.to_string()));
        }
    }

    let Some(agent_id) = connection_config
        .get("agent_id")
        .and_then(serde_json::Value::as_str)
        .filter(|value| !value.trim().is_empty())
    else {
        return Ok(None);
    };

    let agent_id = Uuid::parse_str(agent_id).map_err(|error| error.to_string())?;
    let Some(agent) =
        sqlx::query_as::<_, ConnectorAgent>("SELECT * FROM connector_agents WHERE id = $1")
            .bind(agent_id)
            .fetch_optional(&state.db)
            .await
            .map_err(|error| error.to_string())?
    else {
        return Err(format!("connector agent '{agent_id}' not found"));
    };

    if agent.status != "online" {
        return Err(format!(
            "connector agent '{}' is not online (status: {})",
            agent.name, agent.status
        ));
    }

    let stale_cutoff = Utc::now() - state.agent_stale_after;
    if let Some(last_heartbeat_at) = agent.last_heartbeat_at
        && last_heartbeat_at < stale_cutoff
    {
        return Err(format!(
            "connector agent '{}' heartbeat is stale",
            agent.name
        ));
    }

    Ok(Some(agent.agent_url))
}

File: services/connector-management-service/internal/domain/agent_registry.go
package domain

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/services/connector-management-service/internal/models"
)

type AgentRegistryStore interface {
	GetConnectorAgent(ctx context.Context, id uuid.UUID) (*models.ConnectorAgent, error)
}

type AgentRegistryClock interface {
	Now() time.Time
}

type AgentRegistryResolver struct {
	Store      AgentRegistryStore
	Clock      AgentRegistryClock
	StaleAfter time.Duration
}

func (r AgentRegistryResolver) ResolveAgentURL(ctx context.Context, connectionConfig json.RawMessage) (*string, error) {
	var cfg map[string]any
	_ = json.Unmarshal(connectionConfig, &cfg)
	if url, ok := cfg["agent_url"].(string); ok && strings.TrimSpace(url) != "" {
		trimmed := strings.TrimSpace(url)
		return &trimmed, nil
	}
	rawID, ok := cfg["agent_id"].(string)
	if !ok || strings.TrimSpace(rawID) == "" {
		return nil, nil
	}
	agentID, err := uuid.Parse(strings.TrimSpace(rawID))
	if err != nil {
		return nil, err
	}
	agent, err := r.Store.GetConnectorAgent(ctx, agentID)
	if err != nil {
		return nil, err
	}
	if agent == nil {
		return nil, fmt.Errorf("connector agent '%s' not found", agentID)
	}
	if agent.Status != "online" {
		return nil, fmt.Errorf("connector agent '%s' is not online (status: %s)", agent.Name, agent.Status)
	}
	if r.StaleAfter > 0 && agent.LastHeartbeatAt != nil {
		clock := r.Clock
		if clock == nil {
			clock = realRegistryClock{}
		}
		if agent.LastHeartbeatAt.Before(clock.Now().Add(-r.StaleAfter)) {
			return nil, fmt.Errorf("connector agent '%s' heartbeat is stale", agent.Name)
		}
	}
	url := agent.AgentURL
	return &url, nil
}

type realRegistryClock struct{}

func (realRegistryClock) Now() time.Time { return time.Now().UTC() }
